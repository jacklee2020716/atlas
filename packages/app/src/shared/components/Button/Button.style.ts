import { typography, colors } from '../../theme'
import { makeStyles, StyleFn } from '../../utils'
import { disabled, dimensionsFromProps } from '../../theme/fragments'

export type ButtonStyleProps = {
  text?: string
  type?: 'primary' | 'secondary' | 'tertiary'
  full?: boolean
  size?: 'regular' | 'small' | 'smaller'
  children?: React.ReactNode
  disabled?: boolean
}

const baseStyles: StyleFn = () => ({
  borderWidth: '1px',
  borderStyle: 'solid',
  fontFamily: typography.fonts.headers,
  fontWeight: typography.weights.medium,
  display: 'inline-flex',
  justifyContent: 'center',
  alignItems: 'center',
  color: colors.white,
  '&::selected': {
    background: 'transparent',
  },
})
const colorFromType: StyleFn = (styles = {}, { type }: ButtonStyleProps) => {
  switch (type) {
    case 'primary':
      return {
        ...styles,
        backgroundColor: colors.blue[500],
        borderColor: colors.blue[500],

        '&:hover': {
          backgroundColor: colors.blue[700],
          borderColor: colors.blue[700],
          color: colors.white,
        },
        '&:active': {
          backgroundColor: colors.blue[900],
          borderColor: colors.blue[900],
          color: colors.white,
        },
      }

    case 'secondary':
      return {
        ...styles,
        backgroundColor: colors.black,
        borderColor: colors.blue[500],

        '&:hover': {
          borderColor: colors.blue[700],
          color: colors.blue[300],
        },

        '&:active': {
          borderColor: colors.blue[700],
          color: colors.blue[700],
        },
      }

    case 'tertiary':
      return {
        ...styles,
        backgroundColor: 'transparent',
        borderColor: 'transparent',
        color: colors.blue[500],
        '&:hover': {
          color: colors.blue[300],
        },
        '&:active': {
          color: colors.blue[700],
        },
      }

    default:
      return { ...styles }
  }
}
const paddingFromType: StyleFn = (
  styles,
  {
    size = 'regular',
    children,
    full = false,
  }: { size: 'regular' | 'small' | 'smaller'; children?: React.ReactNode; full: boolean }
) => {
  const buttonHeight = size === 'regular' ? '20px' : size === 'small' ? '15px' : '10px'
  return {
    ...styles,
    padding:
      size === 'regular'
        ? children
          ? '14px 20px'
          : '14px'
        : size === 'small'
        ? children
          ? '12px 14px'
          : '12px'
        : '10px',
    fontSize:
      size === 'regular'
        ? typography.sizes.button.large
        : size === 'small'
        ? typography.sizes.button.medium
        : typography.sizes.button.small,
  }
}

const iconStyles: StyleFn = (styles, { children, size }) => {
  return {
    ...styles,
    marginRight: children != null ? '10px' : '0',
    fontSize:
      size === 'regular'
        ? typography.sizes.icon.large
        : size === 'small'
        ? typography.sizes.icon.medium
        : typography.sizes.icon.small,

    flexShrink: 0,
    '& > *': {
      stroke: 'currentColor',
    },
  }
}

export const useCSS = (props: ButtonStyleProps) => ({
  container: makeStyles([baseStyles, colorFromType, dimensionsFromProps, paddingFromType, disabled])(props),
  icon: makeStyles([iconStyles])(props),
})
