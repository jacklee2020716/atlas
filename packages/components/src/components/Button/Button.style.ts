import { typography, colors } from "../../theme"
import { makeStyles, StyleFn } from "../../utils"
import { disabled, dimensionsFromProps, log } from "../../theme/fragments"

export type ButtonStyleProps = {
	text?: string
	type?: "primary" | "secondary"
	full?: boolean
	size?: "regular" | "small" | "smaller"
	children?: React.ReactNode
	disabled?: boolean
}

const baseStyles: StyleFn = () => ({
	borderWidth: "1px",
	borderStyle: "solid",
	fontFamily: typography.fonts.base,
	fontWeight: typography.weights.medium,
	display: "inline-flex",
	justifyContent: "center",
	alignItems: "center",
	color: colors.white,
	"&::selected": {
		background: "transparent"
	}
})
const colorFromType: StyleFn = (styles, { type }: ButtonStyleProps) => {
	switch (type) {
		case "primary":
			return {
				...styles,
				backgroundColor: colors.blue[500],
				borderColor: colors.blue[500],

				"&:hover": {
					backgroundColor: colors.blue[700],
					borderColor: colors.blue[700],
					color: colors.white
				},
				"&:active": {
					backgroundColor: colors.blue[900],
					borderColor: colors.blue[900],
					color: colors.white
				}
			}

		case "secondary":
			return {
				...styles,
				color: colors.blue[500],
				backgroundColor: colors.black,
				borderColor: colors.blue[500],

				"&:hover": {
					borderColor: colors.blue[700],
					color: colors.blue[300]
				},

				"&:active": {
					borderColor: colors.blue[700],
					color: colors.blue[700]
				}
			}
	}
}
const paddingFromType: StyleFn = (
	styles,
	{
		size = "regular",
		children,
		full
	}: { size: "regular" | "small" | "smaller"; children?: React.ReactNode; full: boolean }
) => {
	const buttonHeight = size === "regular" ? "20px" : size === "small" ? "15px" : "10px"
	return {
		...styles,
		margin: `0 ${full ? "0" : "15px"} 0 0`,
		padding:
			size === "regular"
				? !!children
					? "14px 17px"
					: "14px"
				: size === "small"
				? !!children
					? "12px 14px"
					: "12px"
				: "10px",
		fontSize:
			size === "regular"
				? typography.sizes.button.large
				: size === "small"
				? typography.sizes.button.medium
				: typography.sizes.button.small,

		height: buttonHeight,
		maxHeight: buttonHeight
	}
}

const iconStyles: StyleFn = (styles, { children, size }) => {
	return {
		...styles,
		marginRight: children != null ? "10px" : "0",
		fontSize:
			size === "regular"
				? typography.sizes.icon.large
				: size === "small"
				? typography.sizes.icon.medium
				: typography.sizes.icon.small,

		flexShrink: 0,
		"& > *": {
			stroke: "currentColor"
		}
	}
}

export const useCSS = (props: ButtonStyleProps) => ({
	container: makeStyles([baseStyles, colorFromType, dimensionsFromProps, paddingFromType, disabled])(props),
	icon: makeStyles([iconStyles])(props)
})

// 	text,
// 	type = "primary",
// 	width = "normal",
// 	size = "regular",
// 	disabled = false,
// }: ButtonStyleProps) => {
//

// 	const primaryButton = {
// 		container: css`
// 			border: 1px solid ${colors.blue[500]};
// 			color: ${colors.white};
// 			background-color: ${colors.blue[500]};

// 			padding: ${size === "regular"
// 				? !!text
// 					? "14px 17px"
// 					: "14px"
// 				: size === "small"
// 				? !!text
// 					? "12px 14px"
// 					: "12px"
// 				: "10px"};
// 			font-size: ${size === "regular"
// 				? typography.sizes.button.large
// 				: size === "small"
// 				? typography.sizes.button.medium
// 				: typography.sizes.button.small};
//
//
// 		`,
// 	};

// 	const secondaryButton = {
// 		container: css`
// 			border: 1px solid ${colors.blue[500]};

// 			background-color: ${colors.black};
// 			justify-content: center;
// 			padding: ${size === "regular"
// 				? !!text
// 					? "14px 17px"
// 					: "14px"
// 				: size === "small"
// 				? !!text
// 					? "12px 14px"
// 					: "12px"
// 				: "10px"};
// 			font-size: ${size === "regular"
// 				? typography.sizes.button.large
// 				: size === "small"
// 				? typography.sizes.button.medium
// 				: typography.sizes.button.small};
// 			margin: 0 ${width === "normal" ? "15px" : "0"} 0 0;
// 			height: ${buttonHeight};
// 			max-height: ${buttonHeight};

// 			&:hover {
// 				background-color: ${colors.black};
// 				border-color: ${colors.blue[700]};
// 				color: ${colors.blue[300]};
// 			}

// 			&:active {
// 				background-color: ${colors.black};
// 				border-color: ${colors.blue[700]};
// 				color: ${colors.blue[700]};
// 			}
// 		`,
// 	};

// 	const disabledButton = {
// 		container: css`
// 			border: 1px solid ${colors.white};
// 			color: ${colors.white};
// 			background-color: ${colors.gray[100]};
// 			justify-content: center;
// 			padding: ${size === "regular"
// 				? !!text
// 					? "14px 17px"
// 					: "14px"
// 				: size === "small"
// 				? !!text
// 					? "12px 14px"
// 					: "12px"
// 				: "10px"};
// 			display: ${width === "normal" ? "inline-flex" : "flex"};
// 			align-items: center;
// 			cursor: ${disabled ? "not-allowed" : "default"};
// 			font-family: ${typography.fonts.base};
// 			font-weight: ${typography.weights.medium};
// 			font-size: ${size === "regular"
// 				? typography.sizes.button.large
// 				: size === "small"
// 				? typography.sizes.button.medium
// 				: typography.sizes.button.small};
// 			margin: 0 ${width === "normal" ? "15px" : "0"} 0 0;
// 			height: ${buttonHeight};
// 			max-height: ${buttonHeight};

// 			&:hover {
// 				background-color: ${colors.gray[100]};
// 				border-color: ${colors.white};
// 				color: ${colors.white};
// 			}

// 			&:active {
// 				background-color: ${colors.gray[100]};
// 				border-color: ${colors.white};
// 				color: ${colors.white};
// 			}

// 			&::selection {
// 				background: transparent;
// 			}
// 		`,
// 	};

// 	const icon = css`
// 		margin-right: ${!!text ? "10px" : "0"};
// 		font-size: ${size === "regular"
// 			? typography.sizes.icon.large
// 			: size === "small"
// 			? typography.sizes.icon.medium
// 			: typography.sizes.icon.small};

// 		& > path:nth-of-type(1) {
// 			color: inherit;
// 			flex-shrink: 0;
// 		}
// 	`;

// 	const result = disabled ? disabledButton : type === "primary" ? primaryButton : secondaryButton;
// 	return { icon, ...result };
// };
