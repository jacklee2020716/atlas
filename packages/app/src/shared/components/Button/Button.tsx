import React from "react";
import { SerializedStyles } from "@emotion/core";
import { ButtonStyleProps, useCSS } from "./Button.style";
import { BlockIcon } from "../../icons";

type ButtonProps = {
	children?: React.ReactNode;
	icon?: boolean;
	disabled?: boolean;
	containerCss: SerializedStyles;
	onClick?: (e: React.MouseEvent<HTMLButtonElement>) => void;
} & ButtonStyleProps;

const Button: React.FC<Partial<ButtonProps>> = ({
	children,
	icon,
	type = "primary",
	disabled = false,
	containerCss,
	onClick = () => {},
	...styleProps
}) => {
	let styles = useCSS({ disabled, type, children, ...styleProps });
	return (
		<button css={[styles.container, containerCss]} onClick={onClick} disabled={disabled}>
			{icon && <BlockIcon css={styles.icon} />}
			{children}
		</button>
	);
};

export default Button;
