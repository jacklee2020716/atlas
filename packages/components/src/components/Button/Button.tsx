import React from "react";
import { ButtonStyleProps, useCSS } from "./Button.style";
import BlockIcon from "../../../assets/block.svg";

type ButtonProps = {
	children?: React.ReactNode;
	icon?: boolean;
	disabled?: boolean;
	onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
} & ButtonStyleProps;

export default function Button({
	children,
	icon,
	type = "primary",
	disabled = false,
	onClick,
	...styleProps
}: ButtonProps) {
	let styles = useCSS({ disabled, type, children, ...styleProps });
	return (
		<button css={styles.container} onClick={disabled ? null : onClick}>
			{icon && <BlockIcon css={styles.icon} />}
			{children}
		</button>
	);
}
