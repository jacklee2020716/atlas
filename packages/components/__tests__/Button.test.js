import React from "react"
import { mount } from "enzyme"
import Button from "./../src/components/Button"

describe("Button component", () => {

  it("Should render default button correctly", () => {
    expect(mount(<Button>Click me!</Button>)).toMatchSnapshot()
  })

  it("Should render custom button correctly", () => {
    expect(mount(<Button size="large" color="success">Hello Atlas</Button>)).toMatchSnapshot()
  })

})