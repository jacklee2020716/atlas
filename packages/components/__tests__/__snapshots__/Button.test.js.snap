// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Should render custom button correctly 1`] = `ReactWrapper {}`;

exports[`Button component Should render default button correctly 1`] = `ReactWrapper {}`;
